### PR TITLE
Feature/search room equipment

### PR DESCRIPTION
**Pull request by JelenaPetric00**

### CODE REVIEW

File: src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs [hunk 1]
```diff
@@ -10,5 +15,60 @@ public class RoomEquipmentRepository: BaseRepository<RoomEquipment>, IRoomEquipm
...
+        {
+            foreach (var equipment in roomEquipment.Result)
+            {
+                if (equipment.Quantity >= roomEquipmentDTO.Quantity && equipment.Name.ToLower().Contains(roomEquipmentDTO.Name.ToLower()))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
```
Comment (JanosevicRa177) [inline src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs:72]: there is no database call

File: src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs [hunk 2]
```diff
@@ -10,5 +15,60 @@ public class RoomEquipmentRepository: BaseRepository<RoomEquipment>, IRoomEquipm
...
+
+        public bool checkFloorByEquipmentQuantity(Task<IEnumerable<RoomEquipment>> roomEquipment, RoomEquipmentDTO roomEquipmentDTO)
+        {
+            foreach (var equipment in roomEquipment.Result)
+            {
+                if (equipment.Quantity >= roomEquipmentDTO.Quantity)
+                {
+                    return true;
+                }
```
Comment (JanosevicRa177) [inline src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs:57]: remove braces, unnecessary

Comment (JanosevicRa177) [reply src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs:57]: check in every function

File: src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs [hunk 3]
```diff
@@ -10,5 +15,60 @@ public class RoomEquipmentRepository: BaseRepository<RoomEquipment>, IRoomEquipm
...
+            return await _dataContext.RoomEquipment.Where(r => r.Name.ToLower().Contains(roomEquipmentDTO.Name.ToLower())
+                                                          && r.RoomId.Equals(roomEquipmentDTO.RoomId)).ToListAsync();
+        }
+
+        public async Task<IEnumerable<RoomEquipment>> GetAllByCombineSearchInRoom(RoomEquipmentDTO roomEquipmentDTO)
+        {
+            return await _dataContext.RoomEquipment.Where(r => (r.RoomId == roomEquipmentDTO.RoomId) &&
+                                                                 (r.Name.ToLower().Contains(roomEquipmentDTO.Name.ToLower())) &&
+                                                                 (r.Quantity >= roomEquipmentDTO.Quantity)).ToListAsync();
+        }
```
Comment (JanosevicRa177) [inline src/HospitalLibrary/Rooms/Repositories/RoomEquipmentRepository.cs:36]: you have combined search why do you need another 2?

File: src/HospitalLibrary/Rooms/RoomEquipmentService.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            else if(roomEquipmentDTO.Quantity > 0)
+            {
+                return status + 2;
+            }
+           else
+            {
+                return status + 3;
+            }
+        }
```
Comment (JanosevicRa177) [inline src/HospitalLibrary/Rooms/RoomEquipmentService.cs:80]: why is this necessary?

File: src/HospitalTests/Integrations/Equipment/EquipmentTest.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        public EquipmentTest(TestDatabaseFactory<Startup> factory) : base(factory)
+        {
+        }
+        public RoomsController SetupController(IServiceScope scope)
+        {
+            return new RoomsController(scope.ServiceProvider.GetRequiredService<IRoomService>());
+        }
+
+        [Fact]
```
Comment (JanosevicRa177) [inline src/HospitalTests/Integrations/Equipment/EquipmentTest.cs:27]: needs collection annotation

File: src/HospitalTests/Integrations/Equipment/EquipmentTest.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        public void Room_has_no_equipment()
+        {
+            using var scope = Factory.Services.CreateScope();
+            var controller = SetupController(scope);
+
+            var result = ((OkObjectResult)controller.GetRoomEquipment(5).Result)?.Value as IEnumerable<RoomEquipment>;
+
+            Assert.Empty(result);
```
Comment (JanosevicRa177) [inline src/HospitalTests/Integrations/Equipment/EquipmentTest.cs:47]: use shouldly not assert

File: src/HospitalTests/Integrations/Equipment/EquipmentTest.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+        [Fact]
+        public void Room_has_equipment()
+        {
+            using var scope = Factory.Services.CreateScope();
+            var controller = SetupController(scope);
+
+            var result = ((OkObjectResult)controller.GetRoomEquipment(1).Result)?.Value as IEnumerable<RoomEquipment>;
+
+            Assert.NotEmpty(result);
```
Comment (JanosevicRa177) [inline src/HospitalTests/Integrations/Equipment/EquipmentTest.cs:35]: use shouldly not assert

File: src/HospitalLibrary/Rooms/RoomEquipmentValidator.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        private IUnitOfWork _unitOfWork;
+
+        public RoomEquipmentValidator(IUnitOfWork unitOfWork)
+        {
+            _unitOfWork = unitOfWork;   
+        }
+        public void GetAllByCombineSearchInRoomValidation(RoomEquipmentDTO roomEqipmentDTO)
+        {
+            
+        }
```
Comment (JanosevicRa177) [inline src/HospitalLibrary/Rooms/RoomEquipmentValidator.cs:25]: empty function?

File: src/HospitalLibrary/Rooms/RoomEquipmentService.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                var equipment = _unitOfWork.RoomRepository.GetAllEquipmentbyRoom(room.Id);
+                if (equipment.Result != null)
+                {
+                   if(checkSearchInputFloor(roomEquipmentDTO, equipment))
+                    {
+                        
+                        result.Add(room);
+                    }
+                }
+            }
```
Comment (JanosevicRa177) [inline src/HospitalLibrary/Rooms/RoomEquipmentService.cs:54]: needs to be in floorService, getRoomById and then GetEqupimentOnFloor because room has floor id

**Review by JanosevicRa177 — CHANGES_REQUESTED** [review]

File: src/HospitalAPI/Controllers/Intranet/RoomEquipmentController.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        [HttpGet]
+        [Route("search")]
+        [Authorize(Roles = "Manager")]
+        public async Task<IActionResult> searchEquipmentInRoom([FromQuery] RoomEquipmentDTO roomEquipmentDTO)
+        {
+            var result = await _roomEquipmentService.searchEquipmentInRoom(roomEquipmentDTO);
+            return Ok(result);
+        }
+
+        [Route("floorSearch")]
```
Comment (srdjan1210) [inline src/HospitalAPI/Controllers/Intranet/RoomEquipmentController.cs:29]: This should be floor/search or floor-search

File: src/HospitalAPI/Controllers/Intranet/RoomsController.cs [hunk 11]
```diff
@@ -42,6 +42,15 @@ public IActionResult UpdateName(int id, [FromBody] string name)
             return Ok(room);
         }
 
+        [Route("equipmentView/{roomId:int}")]
```
Comment (srdjan1210) [inline src/HospitalAPI/Controllers/Intranet/RoomsController.cs:45]: If you are getting equipment by room, you should prefix route with room. So instead of equipmentView/{roomId} you should write room/{roomId}, so that the final route becomes api/intranet/equipment/room/1 for example

File: src/HospitalLibrary/Rooms/Repositories/RoomRepository.cs [hunk 12]
```diff
@@ -25,11 +25,15 @@ public async Task<IEnumerable<Room>> FindAllByFloor(int floor)
         public async Task<IEnumerable<Room>> SearchByTypeAndName(RoomSearchDTO roomsSearchDTO, int floorId)
         {
             return await _dataContext.Rooms.Where(r =>
-                ((r.FloorId == floorId))).Where(r=>
-                 (r.RoomNumber.Contains(roomsSearchDTO.RoomName)&&
-                 ((r.RoomType == roomsSearchDTO.RoomType)||(roomsSearchDTO.RoomType==RoomType.NO_TYPE)))
+                ((r.FloorId == floorId))).Where(r =>
```
Comment (srdjan1210) [inline src/HospitalLibrary/Rooms/Repositories/RoomRepository.cs:28]: You don't need parenthesis here, Rooms.Where(r => r.FloorId == floorId) is enough

File: src/HospitalLibrary/Rooms/RoomEquipmentService.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public IEnumerable<Room> searchRoomsByFloorContainingEquipment(RoomEquipmentDTO roomEquipmentDTO)
+        {
+            var rooms = _unitOfWork.RoomEquipmentRepository.CheckFloorByCombineEquipmentSearch(roomEquipmentDTO);
+            List<Room> result = new List<Room>();
+
+            foreach(Room room in rooms)
+            {
+                if (!room.RoomEquipment.Count().Equals(0))
```
Comment (srdjan1210) [inline src/HospitalLibrary/Rooms/RoomEquipmentService.cs:36]: This is ok, but you should refactor this sometime by putting a method inside of room class that will return you boolean if  there is no equipment. For example: if(!room.HasEquipment()), also it's ok to use == instead of Equals for integers.

File: src/HospitalLibrary/Rooms/RoomEquipmentService.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+                    result.Add(room);
+
+            }
+            IEnumerable<Room> resultList = result.AsEnumerable<Room>();
```
Comment (srdjan1210) [inline src/HospitalLibrary/Rooms/RoomEquipmentService.cs:40]: This is excess, you can just say return resul.AsEnumerable<Room>();

**Review by srdjan1210 — CHANGES_REQUESTED** [review]

File: src/HospitalAPI/Controllers/Intranet/RoomEquipmentController.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+            return Ok(result);
+        }
+
+        [Route("floor-search")]
+        [HttpGet]
+        [Authorize(Roles = "Manager")]
+        public IActionResult searchRoomsByFloorContainingEquipment([FromQuery] RoomEquipmentDTO roomEquipmentDTO)
+        {
+            var result = _roomEquipmentService.searchRoomsByFloorContainingEquipment(roomEquipmentDTO);
+            return Ok(result);
```
Comment (srdjan1210) [inline src/HospitalAPI/Controllers/Intranet/RoomEquipmentController.cs:35]: Method names should start with uppercase

**Review by srdjan1210 — CHANGES_REQUESTED** [review]
